### PR TITLE
Add lambda expressions with closures to Zeek.

### PR DESCRIPTION
**Pull request by ZekeMedley**
This allows anonymous functions in Zeek to capture their closures.
they do so by creating a copy of their enclosing frame and joining
that with their own frame.

There is no way to specify what specific items to capture from the
closure like C++, nor is there a nonlocal keyword like Python.
Attemptying to declare a local variable that has already been caught
by the closure will error nicely. At the worst this is an inconvenience
for people who are using lambdas which use the same variable names
as their closures.

As a result of functions copying their enclosing frames there is no
way for a function with a closure to reach back up and modify the
state of the frame that it was created in. This lets functions that
generate functions work as expected. The function can reach back 
and modify its copy of the frame that it is captured in though.

Implementation wise this is done by creating two new subclasses in
Zeek. The first is a LambdaExpression which can be thought of as a
function generator. It gathers all of the ingredients for a function
at parse time, and then when evaluated creats a new version of that
function with the frame it is being evaluated in as a closure. The
second subclass is a ClosureFrame. This acts for most intents and
purposes like a regular Frame, but it routes lookups of values to its
closure as needed.

### CODE REVIEW

**Comment by ZekeMedley** [comment]
Some thoughts on this that I'd be interested in hearing what people think:

1. Right now this does not apply to functions that give the default values of a table. Without too much effort it could - just a matter of changing how its parsed. I tend to think that those functions should be included because I could see it being pretty useful. Any thoughts on rather or not we should add that?

2. The second non-merge commit changes how frames are cloned such that they only clone elements that are found by `OuterIDBindingFinder`. The reason I think doing that might be useful is that for large closures with expensive to clone elements (ex: a big paraglob) the default behaviour of cloning the entire closure could add a pretty substantial performance overhead. This does seem less "safe" though. I'd really appreciate some review of those methods because even worse than a performance overhead would be trying to look up something in the closure that wasn't cloned and getting a seg-fault.

3. More generally, regarding performance, does anyone have any thoughts on the performance impact of this? This does add some overhead to things. Capturing a closure is potentially an expensive operation and ClosureFrame's do a little more work than a regular Frame to perform a lookup because they have to route it to one of their subframes.

File: src/Frame.cc [hunk 1]
```diff
@@ -109,3 +169,121 @@ void Frame::ClearTrigger()
...
+	{
+	Frame* new_closure = this->closure->Clone();
+	Frame* new_regular = this->body->Clone();
+
+	ClosureFrame* cf = new ClosureFrame(new_closure, new_regular, nullptr);
+	cf->closure_elements = this->closure_elements;
+	return cf;
+	}
+
+Frame* ClosureFrame::SelectiveClone(id_list* choose)
```
Comment (ZekeMedley) [inline src/Frame.cc]: This is where selective cloning takes place and the part of this that could probably use the most thought. As far as I can think (assuming that everything is implementing traverse properly) [`OuterIDBindingFinder`](https://github.com/zeek/zeek/blob/7cf623b2ff2288ea01059b2a5ed4712e112f0526/src/Var.cc#L405) should catch everything that we need to clone, but I'm not too familiar with the corners of Zeek. I'd appreciate a second brain here.

Comment (ZekeMedley) [reply src/Frame.cc]: Selective cloning also happens in `Frame::SelectiveClone`. This is the more interesting of the two methods though.

File: src/Frame.cc [hunk 2]
```diff
@@ -27,12 +29,56 @@ Frame::Frame(int arg_size, const BroFunc* func, const val_list* fn_args)
...
+	this->break_before_next_stmt = false;
+	this->break_on_return = false;
+	this->delayed = false;
+
+	if ( other->trigger )
+		Ref(other->trigger);
+
+	for (int i = 0; i < size; i++)
+	  if (frame[i])
+	    Ref(frame[i]);
```
Comment (ZekeMedley) [inline src/Frame.cc]: @jsiwek I could use a sanity check on reference counting here if you get a chance. 

This constructor is intended to create a "view" of the input frame. ClosureFrame uses it so that it can route most methods to the underlying frame without having to add a decent amount of extra code to do that. I'm Refing everything I'm viewing because when ClosureFrame is taken apart it Unrefs it all. I wasn't getting any trouble without the Ref calls on all the elements though from my tests so I wonder if it is actually needed.

File: src/parse.y [hunk 3]
```diff
@@ -1216,16 +1216,42 @@ func_body:
 	;
 
 anonymous_function:
-		TOK_FUNCTION begin_func func_body
-			{ $$ = new ConstExpr($2->ID_Val()); }
+		TOK_FUNCTION func_params
+			{
+			$<id>$ = current_scope()->GenerateTemporary("lambda-function");
+			begin_func($<id>$, current_module.c_str(), FUNC_FLAVOR_FUNCTION, 0, $2);
+			}
```
Comment (ZekeMedley) [inline src/parse.y]: This could call `begin_func` here. I decided not to though because unless we decide to make all anonymous functions capture their scope (specifically ones the are the default for tables) calling these functions `lambda-function` adds some clarity at the cost of two lines of mostly redundant code.

**Comment by jsiwek** [comment]
> * Right now this does not apply to functions that give the default values of a table. Without too much effort it could 

Seems worth adding support for that if it's fairly trivial, if only for sake of consistency.

> * The second non-merge commit changes how frames are cloned such that they only clone elements that are found by `OuterIDBindingFinder`.

That sounds Good to me, no reason to clone other things that we know aren't referenced in the closure.  Don't think there's a safety issue at the conceptual level -- if somehow the closure does look up something that didn't get cloned, that means there's a bug in how we implemented the non-local name reference lookup/resolver and need to fix it.

> * More generally, regarding performance, does anyone have any thoughts on the performance impact of this?

If performance overhead is isolated to closure usage, that seems fine to me (suspect closures won't be used at a frequency anywhere near normal functions).  Is there any overhead added to function usage (non-closures) in general ?

**Comment by ZekeMedley** [comment]
> If performance overhead is isolated to closure usage, that seems fine to me (suspect closures won't be used at a frequency anywhere near normal functions). Is there any overhead added to function usage (non-closures) in general ?

There shouldn't be any changes to how normal functions work with this and BroFuncs that don't have a closure will just use a normal Frame so they won't see any of the extra time added by lookups in the `std::unordered_set`. I'll see about writing some benchmarking script so that I can put numbers to that.

File: src/Func.cc [hunk 4]
```diff
@@ -127,6 +127,15 @@ void Func::AddBody(Stmt* /* new_body */, id_list* /* new_inits */,
 	}
 
 
+Val* Func::DoClone()
+	{
+	// By default, ok just to return a reference. Func does not have any "state".
+	// That is different across instances.
+	Val* v = new Val(this);
+	Ref(v);
```
Comment (rsmmr) [inline src/Func.cc]: Don't need this one. it's already +1 from the new.

File: src/Func.cc [hunk 5]
```diff
@@ -454,6 +474,49 @@ void BroFunc::AddBody(Stmt* new_body, id_list* new_inits, int new_frame_size,
...
+	// store a copy of this closure.
+	if ( ! this->closure )
+		{
+		return Func::DoClone();
+		}
+	else
+		{
+		BroFunc* other = new BroFunc();
+
+		other->bodies = this->bodies;
```
Comment (rsmmr) [inline src/Func.cc]: These need to be ref'ed, the dtor unrefs them.

File: src/Func.cc [hunk 6]
```diff
@@ -454,6 +474,49 @@ void BroFunc::AddBody(Stmt* new_body, id_list* new_inits, int new_frame_size,
...
+		other->scope = this->scope;
+		other->kind = this->kind;
+		other->type = this->type;
```
Comment (rsmmr) [inline src/Func.cc]: It isn't so nice to clone the member variables of the parent class here. If the parent class ever changes, it's likely that one would forget to update this code. I realize that calling `Func::DoClone()` won't work, as it optimizes for the cases that it doesn't need to maintain state. But how about at least adding a protected method `CopyState()` or so to `Func` that takes care of the copying its state over into a new instance. That way the logic is associated with that class at least.

Comment (rsmmr) [reply src/Func.cc]: 2nd note is here is that `Func::~Func` unrefs `type`, so it needs to be ref'ed on clone.

File: src/Func.cc [hunk 7]
```diff
@@ -473,7 +536,8 @@ Stmt* BroFunc::AddInits(Stmt* body, id_list* inits)
 		return body;
 
 	StmtList* stmt_series = new StmtList;
-	stmt_series->Stmts().append(new InitStmt(inits));
+	InitStmt* first = new InitStmt(inits);
```
Comment (rsmmr) [inline src/Func.cc]: why this change?

File: src/Func.h [hunk 8]
```diff
@@ -17,6 +20,8 @@ class Frame;
 class ID;
 class CallExpr;
 
+struct CloneState;
```
Comment (rsmmr) [inline src/Func.h]: Do we need this?

File: src/Func.h [hunk 9]
```diff
@@ -59,6 +64,8 @@ class Func : public BroObj {
 	void Describe(ODesc* d) const override = 0;
 	virtual void DescribeDebug(ODesc* d, const val_list* args) const;
 
+	virtual Val* DoClone();
```
Comment (rsmmr) [inline src/Func.h]: It looks like you're implementing the API from `Val` here, which I think isn't really necessary, as `Func` is outside of that hierarchy. I'd change this to `virtual Func* Clone()`, and then override that in derived classes.

File: src/Func.h [hunk 10]
```diff
@@ -89,9 +96,14 @@ class BroFunc : public Func {
 	int IsPure() const override;
 	Val* Call(val_list* args, Frame* parent) const override;
 
+        void AddClosure(std::shared_ptr<id_list> ids, Frame* f);
 	void AddBody(Stmt* new_body, id_list* new_inits, int new_frame_size,
 			int priority) override;
 
+	void fsets();
```
Comment (rsmmr) [inline src/Func.h]: What's this?

File: src/parse.y (hunk 3 above)
Comment (rsmmr) [reply src/parse.y]: Agree with Jon here that we should cover all cases for consistency.

File: src/Func.h [hunk 11]
```diff
@@ -136,6 +165,18 @@ struct CallInfo {
 	const val_list* args;
 };
 
+// Struct that collects the arguments for a Func.
+// Used for BroFuncs with closures.
+struct function_ingredients
```
Comment (rsmmr) [inline src/Func.h]: I haven't traced this through: is the reference counting done where necessary when storing pieces in these ingredients? Alternatively, the struct could maintain ref counts itself for the duration of an instance's lifetime.

File: src/Expr.cc [hunk 12]
```diff
@@ -4319,6 +4320,52 @@ void CallExpr::ExprDescribe(ODesc* d) const
...
+	lamb->AddClosure(outer_ids, f);
+
+	ingredients->id->SetVal((new Val(lamb))->Ref());
+	ingredients->id->SetConst();
+	ingredients->id->ID_Val()->AsFunc()->SetScope(ingredients->scope);
+
+	return ingredients->id->ID_Val();
+	}
+
+void LambdaExpr::ExprDescribe(ODesc* d) const
```
Comment (rsmmr) [inline src/Expr.cc]: This should render back Bro script code, at least when in "readable" mode. Take a look at some of the other `ExprDescribe` methods.

File: src/Expr.cc [hunk 13]
```diff
@@ -4319,6 +4320,52 @@ void CallExpr::ExprDescribe(ODesc* d) const
...
+	BroFunc* lamb = new BroFunc(
+		ingredients->id,
+		ingredients->body,
+		ingredients->inits,
+		ingredients->frame_size,
+		ingredients->priority);
+
+	lamb->AddClosure(outer_ids, f);
+
+	ingredients->id->SetVal((new Val(lamb))->Ref());
```
Comment (rsmmr) [inline src/Expr.cc]: `new' already refs.

Comment (rsmmr) [reply src/Expr.cc]: A question: Do we need to set the ID's value here, rather than just passing the new value back directly? I'm not quite sure but something here doesn't feel quite right: we keep modifying the value of the ID each time the lambda function gets executed. That means one can't really rely on the ID's value, and it's a temporary ID that's not visible to a script anyways.

File: src/Frame.h [hunk 14]
```diff
@@ -4,25 +4,31 @@
 #define frame_h
 
 #include <vector>
-using namespace std;
+#include <unordered_set>
+#include <memory>
 
 #include "Val.h"
 
+using namespace std;
```
Comment (rsmmr) [inline src/Frame.h]: Please don't use "using namespace" in a header file, it'll propagate to everybody including it.

File: src/Frame.h [hunk 15]
```diff
@@ -82,6 +90,65 @@ class Frame : public BroObj {
...
+	Frame* closure;
+	Frame* body;
+
+	// Searches this frame and all sub-frame's closures for a value corresponding
+	// to the id.
+        static Val* GatherFromClosure(const Frame* start, const ID* id);
+        // Moves through the closure frames and associates val with id.
+        static void SetInClosure(Frame* start, const ID* id, Val* val);
+
+        bool closure_contains(const ID* i) const
```
Comment (rsmmr) [inline src/Frame.h]: Use CamelCaps for all methods.

File: src/Frame.h [hunk 16]
```diff
@@ -82,6 +90,65 @@ class Frame : public BroObj {
...
+
+			return h;
+			}
+	};
+
+	// NOTE: In a perfect world this would be best done with a trie or bloom
+	// filter. We only need to check if things are NOT in the closure.
+        // In reality though the size of a closure is small enough that operatons are
+        // fairly quick anyway.
+	std::unordered_set<const char*, const_char_hasher> closure_elements;
```
Comment (rsmmr) [inline src/Frame.h]: In fact the set will typically be tiny, meaning a vector with linear search might be the most efficient thing.

**Comment by rsmmr** [comment]
Very cool. I haven't fully looked through yet, but left a 1st set of comments from a quick skimming over.

**Comment by ZekeMedley** [comment]
Thanks so much for taking the time to give this a look. These are all very helpful comments. I'll get to this shortly.

File: src/Expr.h [hunk 17]
```diff
@@ -937,14 +943,40 @@ class CallExpr : public Expr {
...
+/*
+	Class to handle the creation of anonymous functions with closures.
+
+	Facts:
+		- LambdaExpr creates a new BroFunc on every call to Eval.
+		- LambdaExpr must be given all the information to create a BroFunc on
+		  construction except for the closure.
+		- The closure for created BroFuncs is the frame that the LambdaExpr is
+		  evaluated in.
+*/
```
Comment (0xxon) [inline src/Expr.h]: Really minor nit - the comment formatting does not match with how we typically format comments.

File: src/Expr.cc (hunk 13 above)
Comment (ZekeMedley) [reply src/Expr.cc]: I think this comes down to the principal of least astonishment. Will people expect that the id has a value equal to the last function that it generated, or that it has none at all? Now that I think about it more, and given that it threw you a little, I figure it would probably be best just not to give this id a value. The alternative is that it has a value equal to the last function it generated, but I'm not sure that is useful, nor is it particularly intuitive.

File: src/Expr.h (hunk 17 above)
Comment (ZekeMedley) [reply src/Expr.h]: I tried to clean this up a little, although I am a little unclear on what proper comment formatting is for long comments like that is. Seemed like there is pretty minimal usage of multi-line comments, so I went ahead and replaced it with single line ones. Let me know if that's the wrong style as well.

File: src/Frame.h (hunk 16 above)
Comment (ZekeMedley) [reply src/Frame.h]: Good point :) I updated that.

File: src/Expr.h (hunk 17 above)
Comment (rsmmr) [reply src/Expr.h]: Generally Doxygen style is what we've been using for recent code.

**Comment by ZekeMedley** [comment]
How to we expect functions with closures being serialized over Broker to behave?

Right now, sending a function over broker just instructs the other side to use its version of that function. For example, given two function declarations like these in the sending and receiving processes:
```zeek
# sender foo
function foo (bar: count)
    {
    print bar;
    }
```
```zeek
# receiver foo
function foo (bar: count)
    {
    print bar * 3;
    }
```
If the sender sends `foo` to the receiver in an event and the receiver calls `foo(3)` they will get `9`.

When dealing with functions with closures it seems like the question is rather or not we'd like to use the sent closure, or the local one? We're guaranteed that the closure exists on the receiving side because otherwise the function declaration wouldn't be valid.

It seems to me that given how function serialization works now, we should just continue to use the local version of the function, rather than using the sent closure. If someone wants to update values in the receivers closure they could still send those over and set them manually. Any thoughts? @jsiwek @rsmmr

**Comment by rsmmr** [comment]
@ZekeMedley and I just talked: using the local function with  the original closure values seems the right thing to do, plus however catching if the local function is actually different from the remote one to avoid executing unexpected code.

**Comment by rsmmr** [comment]
@zeke What's the state here, is it ready for another look>

**Comment by ZekeMedley** [comment]
Coming along :) I'll clean up those commits as soon as I've got everything working nicely. Sorry this is taking a while. I was sick a couple days last week and sending functions over broker is proving to be a little non-trivial.

As it stands in those commits I just pushed: I have anonymous functions sending over broker and verifying that the function on the receiving end has the same body fine now. I also have closure serialization code that is mostly working for the closures. Some more recursive magic in there. 

Some kinks that need ironing out still:
1. Functions don't currently capture global variables in their closures. If they did, functions wouldn't be able to modify global state and I don't think that's what we want. This gets a little strange though when sending a function over broker: I don't think its super intuitive that the function that the receiver gets will use the receiver's global variables but the sender's captured variables.

2. We require that the receiver has already declared the function that is being sent on their end and that it has the same body. It's not intuitive though where the receiver should have to do this. Should the function have to be declared in the global scope? (This won't work because you can't capture a closure in the global scope) The scope that the function is being sent into? In my current draft, I let the receiver declare them anywhere. I'm not super happy with this though. Again, I'm not sure its very intuitive what the answer should be.

In summary, not being able to send the body of a function over broker (without writing serialization code for Zeek's statements) leads to some behaviour that, imo, isn't particularly intuitive; the fact that function's don't capture global variables leads to weird blending between what information comes from the sent closure, and what information comes from the receiving end; and, as a result of the first point, the fact that the receiver has to have an identical declaration of the function on their end makes for some more less than intuitive behaviour surrounding where that declaration should be.

I'm beginning to think now that the best way to go forward will be to allow anonymous functions to be sent over broker, but until we can serialize the body of a function use the closure that exists locally. Thoughts?

**Comment by ZekeMedley** [comment]
This is ready for another look now :)

**Comment by ZekeMedley** [comment]
I found a reference counting issue when capturing functions in closures this morning. Looks like there is slightly different handling of when functions are getting unreferenced as calling bark or eat below will raise an error. I'll push another commit when this is fixed.
```zeek
function dog_maker(name: string, weight: count) : function (action: string)
	{
	
	local bark = function (who: string) { print fmt("bark i am %s", who); };
	local eat = function (lbs: count) { print fmt("eat i weigh %s", lbs); };

	return function (action: string)
		{
		switch action
			{
			case "bark":
				bark(name);
				break;
			case "eat":
				++weight;
				eat(weight);
				break;
			default:
				print "im not sure i understand";
				break;
			}
		};
	}
```

**Comment by rsmmr** [comment]
I made a 1st pass over the branch and pushed my changes into a separate branch topic/robin/closure-merge. Please continue from that one. 

This is good stuff. Most of my changes are smaller cleanup/coding style/etc changes. Let me know if you have any questions. Further requests/questions:

- I left a few "TODO(robin)" in the code. Please take a look at those. Once of them is unifying the frames, as we discussed.
- Have you tried using "when" statements inside functions with closures? Would be good to confirm they work as expected, as they are messing with frames as well ...
- Have you run leak checks against the new functionality yet? Some tests for that would be good.

I'll then make a 2nd pass later then.

**Comment by ZekeMedley** [comment]
Great. Thanks for taking a look :)

I think you're right about unifying the frames. I sketched it out on paper and with `Frame` having a offset map for serialization it makes sense that they just be unified. Most of the methods will transfer over quite nicely and `GatherFromClosure` will be able to avoid the `dynamic_cast`.

I'll add tests for when statements and leak checks as well. I skimmed the when statement while writing this and it didn't raise any red flags, but I will add a test for that.

**Comment by ZekeMedley** [comment]
Its possible that there is a unit test that I missed that uses the names of functions, but if not I think this might be finally read for a merge :) I'm working on a blog post with Amber about it right now which we'll try and get up shortly after this is merged as there are a lot of interesting parts.

This final clean-up does the following:

- Combines `ClosureFrame` and `Frame` to make the code more readable.
- Changes all of the copy/ serialization code to reflect that.
- Adds a leak test and quite a few additional regular tests.
- Makes functions capture their closures by reference. This means that they can modify the items in the closure they capture. This adds a ton of functionality to function's with closures.
- Adds method descriptions to `frame.h`.

The sheer number of changes to the frame class in this commit makes the diff hard to read. For reviewers, I'd recommend just looking at files changed rather than individual commits because this pull request has evolved a lot since last month when I opened it.

**Comment by ZekeMedley** [comment]
There seems to be some synchronization issues between Travis and GitHub, but [according to Travis](https://travis-ci.com/zeek/zeek/builds/120769991) this build passed fine.
